strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/npm/NpmPackageController.java

```diff
@@ -0,0 +1,313 @@
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
+import org.apache.commons.io.IOUtils;
+import org.carlspring.strongbox.artifact.coordinates.NpmArtifactCoordinates;
+import org.carlspring.strongbox.controllers.BaseArtifactController;
+import org.carlspring.strongbox.npm.metadata.Package;
+import org.carlspring.strongbox.providers.ProviderImplementationException;
+import org.carlspring.strongbox.providers.io.RepositoryPath;
+import org.carlspring.strongbox.providers.layout.NpmLayoutProvider;
+import org.carlspring.strongbox.services.ArtifactManagementService;
+import org.carlspring.strongbox.storage.repository.Repository;
+import org.javatuples.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.util.Assert;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * This Controller used to handle npm requests.
+ * 
+ * @author Sergey Bespalov
+ *
+ */
+@RestController
+@RequestMapping(path = NpmPackageController.ROOT_CONTEXT, headers = "user-agent=npm/*")
+public class NpmPackageController extends BaseArtifactController
+{
+
+    private static final String FIELD_NAME_LENGTH = "length";
+
+    private static final String FIELD_NAME_ATTACHMENTS = "_attachments";
+
+    private static final String FIELD_NAME_VERSION = "versions";
+
+    private static final Logger logger = LoggerFactory.getLogger(NpmPackageController.class);
+
+    public final static String ROOT_CONTEXT = "/storages";
+
+    @Inject
+    private ArtifactManagementService npmArtifactManagementService;
+
+    @Inject
+    private NpmLayoutProvider npmLayoutProvider;
+
+    @Inject
+    @Qualifier("npmJackasonMapper")
+    private ObjectMapper npmJackasonMapper;
+
+    @PreAuthorize("hasAuthority('ARTIFACTS_DEPLOY')")
+    @RequestMapping(path = "{storageId}/{repositoryId}/{name:.+}", method = RequestMethod.PUT, consumes = MediaType.APPLICATION_JSON)
+    public ResponseEntity publish(@PathVariable(name = "storageId") String storageId,
+                                  @PathVariable(name = "repositoryId") String repositoryId,
+                                  @PathVariable(name = "name") String name,
```

Comment:
(For my understanding) : Is this supposed to be `name`, or should it be `path`?

Reply:
`name` is how it named in npm. [see here](https://docs.npmjs.com/files/package.json#name)

Reply:
Okay, thanks for explaining! :)